llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -618,11 +619,40 @@ class YkIRWriter {
 
   void serialiseBlock(BasicBlock &BB, ValueLoweringMap &VLMap,
                       unsigned &BBIdx) {
+    // Keep the instruction skipping logic in one place.
+    auto ShouldSkipInstr = [](Instruction *I) {
+      // Skip non-semantic instrucitons for now.
+      //
+      // We may come back to them later if we need better debugging
+      // facilities, but for now they just clutter up our AOT module.
+      return I->isDebugOrPseudoInst();
+    };
+
+    // Count instructions.
+    //
+    // FIXME: I don't like this much:
+    //
+    //  - Assumes one LLVM instruction becomes exactly one Yk IR instruction.
+    //  - Requires a second loop to count ahead of time.
+    //
+    // Can we emit the instrucitons into a temp buffer and keep a running count
+    // of how many instructions we generated instead?
+    size_t NumInstrs = 0;
+    for (Instruction &I : BB) {
+      if (ShouldSkipInstr(&I)) {
+        continue;
+      }
+      NumInstrs++;
+    }
+
     // num_instrs:
-    OutStreamer.emitSizeT(BB.size());
+    OutStreamer.emitSizeT(NumInstrs);
     // instrs:
     unsigned InstIdx = 0;
     for (Instruction &I : BB) {
+      if (ShouldSkipInstr(&I)) {
```

Comment:
Doesn't this invalidate the "one LLVM instruction == one yk IR instruction" assumption?

Reply:
We are fine for now.

Before this change we counted the number of LLVM AOT instructions and assumed that we would see as many instructions in our AOT IR.

When I initially skipped the debug instructions, I broke the serialiser, because by skipping debug instructions, the count was too high, and I got UB fun at runtime.

The `ShouldSkipInstr` check ensures the count is valid again.

If later any instruction lowering emits either 0 or >1 Yk instruction, we will have to hack around that too.

I'd like it if we could serialise the instructions to a buffer and count as we go, but that's a bigger change that I don't want to get into now.

Reply:
All of this to say: I've not introduced any new badness, only hacked around existing badness :)